crates/telio-nurse/src/heartbeat.rs

```diff
@@ -954,15 +918,194 @@ impl Analytics {
+            let mut writer = WriterBuilder::new()
+                .flexible(true)
+                .delimiter(b':')
+                .terminator(csv::Terminator::Any(b','))
+                .from_writer(vec![]);
+
+            writer.serialize((
+                format!(
+                    "{:x}",
+                    md5::compute(r_segment.server_address.to_string().as_bytes())
+                ),
+                &r_segment.duration.as_secs(),
+                r_segment.connection_data,
+            ))?;
+            relay_str.push_str(String::from_utf8(writer.into_inner()?)?.as_str());
+        }
+
+        // Pop last ',' from 'peers_str' and 'relay_str
+        let _ = peers_str.pop();
+        let _ = relay_str.pop();
+
+        hb_info.nat_traversal_conn_info = peers_str.clone();
+        hb_info.derp_conn_info = relay_str.clone();
+
+        Ok(())
+    }
+
+    async fn get_current_meshmap_info(&mut self) -> HeartbeatMeshmapInfo {
+        // Add our temporary meshnet_id to collection before we try to resolve it
+        self.collection
+            .add_meshnet_id(self.public_key, self.meshnet_id);
+        let config_local_nodes = self
+            .config_nodes
+            .iter()
+            .filter(|(_, is_local)| **is_local)
+            .map(|(pk, _)| *pk)
+            .collect::<HashSet<_>>();
+        self.meshnet_id = self
+            .collection
+            .resolve_current_meshnet_id(&config_local_nodes);
+
+        // All local nodes will have the same meshnet id
+        for pk in config_local_nodes.iter() {
+            if let NodeInfo::Node { meshnet_id, .. } = self.local_nodes.entry(*pk).or_default() {
+                *meshnet_id = Some(self.meshnet_id);
+            }
+        }
+        self.collection
+            .fingerprints
+            .insert(self.public_key, self.config.fingerprint.clone());
+
+        // Insert all the nodes from the config
+        for (node, _) in self.config_nodes.iter() {
+            self.collection
+                .fingerprints
+                .entry(*node)
+                .or_insert_with(|| String::from("null"));
+        }
+
+        // Update stored fingerprints
+        for (node, fp) in self.collection.fingerprints.iter() {
+            self.fingerprints
+                .entry(*node)
+                .and_modify(|stored_fp| {
+                    if *fp != "null" {
+                        *stored_fp = fp.clone()
+                    }
+                })
+                .or_insert_with(|| fp.clone());
+        }
+
+        let nodes_to_retain: HashSet<_> = self.collection.fingerprints.keys().collect();
+        self.fingerprints
+            .retain(|pk, _| nodes_to_retain.contains(pk));
+
+        // Use BTreeSet to sort out the public keys
+        // We sort public keys, instead of sorting by fingerprints, since fingerprints can be empty or null, resulting in
+        // multiple same values, which make having a consistent layout for each node awkward to implement
```

Comment:
Hmm, I don't understand this part tbh, if fingerprint is just some hash of the key (or was it sth else?), it shouldn't be null. Also, if it would be empty it seems that it won't be in the right place for the corresponding index in the following part of the code.

Reply:
Tbh, I just moved this part from `handle_aggregation` and didn't want to break any tests or heartbeat functionality as a whole.

Reply:
It's good to know what the code does when you move it ^^'